README.md
# howsweather
[![Build Status](https://travis-ci.org/PugliaSOS/howsweather.svg?branch=master)](https://travis-ci.org/PugliaSOS/howsweather)
[![codecov](https://codecov.io/gh/PugliaSOS/howsweather/branch/master/graph/badge.svg)](https://codecov.io/gh/PugliaSOS/howsweather)

A simple Node.js script that displays weather on terminal.

## Installation

Make sure you've Node.js `>= 6.3.1` and `npm` installed on your system.

Then type:

```shell
npm install howsweather -g
```

Here you are! You're ready to use **howsweather**.

## Getting Started
![Screenshot](http://i.imgur.com/GO0K4D6.png "Screenshot")

### Configure the necessary

In order to work, **howsweather** reads configuration options from a 
JSON file `.howsweather` in your `$HOME` directory.

So open your favourite text editor and type:
```json
{
  "city": "YOURCITY",
  "api_key": "YOURAPIKEY"
}
```

If you don't specify a city, it'll find your location automatically.

You can get an API key at [OpenWeatherMap](https://openweathermap.org/api).

#### Scales

_howsweather_ allows you to show temperatures in the three main scales.
By default, it uses your position in order to understand what scale use.

You can change this behavior specifying a `scale` option in your 
configuration file, there are three possible values:

- `C` for Celsius scale
- `F` for Fahrenheit scale
- `K` for Kelvin scale

### See weather stats

If you want to see weather for a specific city:
```shell
howsweather "New York"
```

Or for the default one:
```shell
howsweather
```

## License

howsweather is released under terms of [GNU/GPL 3.0 license](LICENSE).


CHANGELOG.md
## 0.1.4 - January 23th, 2017
### Added
- The script use the right temperature scale looking at user's country.
- Unicode emoji for various weather's states: clouds, rain and snow

### Changes
- Clear code adding ESLint
- Unit testing using Mocha and Chai


test/ipinfo.spec.js
const expect = require('chai').expect;

const ipinfo = require('./../src/ipinfo');

describe('IPInfo', () => {
  describe ('#get()', () => {
    it('should returns a Promise', () => {
      expect(ipinfo.get().toString()).to.equal('[object Promise]');
    });
  });
});


test/utilities.spec.js
const expect = require('chai').expect;

const $ = require('./../src/utilities');

describe('Utilities', function () {
  describe('#findCountryScale(String)', function () {
    it('should returns country temperature scale', function () {
      expect($.findCountryScale('IT')).to.equal('C');
      expect($.findCountryScale('US')).to.equal('F');
      expect($.findCountryScale('bz')).to.equal('F');
    });
  });

  describe('#appendEmoji(Object)', function () {
    const getEmoji = $.appendEmoji({ censored: 'X', smile: ':)' });

    it('should append the right emoji to its keyword', function () {
      expect(getEmoji('Smile')).to.equal(':) Smile');
    });

    it('should silently continue if no emoji is found', function () {
      expect(getEmoji('Unknown')).to.equal(' Unknown');
    });
  });

  describe('#formatWeather(Object, String, String)', function () {
    it('should format weather data following a certain schema', function () {
      const fake_temperature = { min: '2', max: '3'};
      const expected = 'MIN 2°C, MAX 3°C | Rain';

      expect($.formatWeather(fake_temperature, 'C', 'Rain')).to.equal(expected);
    });
  });
});


test/openweather.spec.js
const expect = require('chai').expect;

const OpenWeather = require('./../src/openweather');

const weather = new OpenWeather('', 'C');

describe('OpenWeather', function() {
  describe('#setAPI()', function() {
    it('should edits API key', function() {
      weather.setAPI('MOCHA_TEST');
      expect(weather.apiKey).to.equal('MOCHA_TEST');
    });
  });

  describe('#setScale()', function() {
    it('should raises an error if the scale is not valid', function() {
      expect(function() {
        weather.setScale(0);
      }).to.throw(Error);
    })

    it ('should edits scale', function() {
      weather.setScale('F');
      expect(weather.scale).to.equal('F');
    });
  })

  describe('#convertToUserScale()', function() {
    it('should convert from Kelvin to Celsius', function() {
      weather.setScale('C');
      expect(weather.convertToUserScale(1)).to.equal('-272.15');
    });

    it('should convert from Kelvin to Fahrenheit', function() {
      weather.setScale('F');
      expect(weather.convertToUserScale(1)).to.equal('-457.87');
    })
  });
});


src/utilities.js
// Groups together utilities' functions used across the program.

/*
 * Find the temperature's scale used in a certain country,
 * by passing it country's ISO code.
 * ```
 * findCountryScale('IT') //=> C
 * findCountryScale('US') //=> F
 * ```
 */
const findCountryScale = (country) => {
  const targetCountry = country.toLowerCase();

  if (['us', 'bs', 'bz', 'ky', 'pw'].includes(targetCountry)) {
    return 'F';
  }

  return 'C';
};

/*
 * Append emoji to its text representation.
 * Its first argument is the object which contains emojis.
 * ```
 * const emojis = { rain: '\u2614' };
 * const getWeatherEmoji = appendEmoji(emojis);
 *
 * getWeatherEmoji('Rain');
 * ```
 */
const appendEmoji = emoji => (keyword) => {
  const targetKeyword = keyword.toLowerCase();
  const targetEmoji = emoji[targetKeyword] || '';

  return targetEmoji.concat(' ', keyword);
};

/*
 * Returns information about weather in a human-readable way:
 * ```
 *  MIN 2°C, MAX 5°C | Clouds
 * ```
 */
const formatWeather = (data, scale, status) => {
  const suffix = '°'.concat(scale);
  return `MIN ${data.min.concat(suffix)}, MAX ${data.max.concat(suffix)} | ${status}`;
};

module.exports = { findCountryScale, appendEmoji, formatWeather };


src/openweather.js
// Access to weather stats using OpenWeatherMap APIs.
const http = require('http');

const endpoint = 'http://api.openweathermap.org/data/2.5/weather?q={city}&apikey={api}';

/*
 * This module allows you to easily retrieve weather using OpenWeather API.
 * All you need is an API key you can get here: https://openweathermap.org/api.
 *
 *   @example
 *     let weather = new OpenWeather();
 *     weather.setAPI('YOUR_API_KEY')
 *     weather.setScale('C');
 *     weather.on('New York');
 */
function OpenWeather(apiKey, scale) {
  this.setAPI(apiKey);
  this.setScale(scale);
  return this;
}

OpenWeather.prototype.setAPI = function setAPI(key) {
  this.apiKey = key;
};

OpenWeather.prototype.setScale = function setScale(givenScale) {
  if (!['C', 'F', 'K'].includes(givenScale)) {
    throw new Error("Invalid scale for 'setScale' , expected: C, F or K.");
  }

  this.scale = givenScale;
};

/* Get weather for a specific city. */
OpenWeather.prototype.on = function on(city, cb) {
  const self = this;
  const requestURL = this.composeURL(city);

  http.get(requestURL, (res) => {
    res.on('data', self.formatWeather.bind(self, cb));
    res.on('error', console.log);
  });
};

OpenWeather.prototype.formatWeather = function formatWeather(cb, rawData) {
  const data = JSON.parse(rawData);

  cb({
    min: this.convertToUserScale(data.main.temp_min),
    max: this.convertToUserScale(data.main.temp_max),
    weatherStatus: data.weather[0].main
  });
};

OpenWeather.prototype.convertToUserScale = function convertToUserScale(temperature) {
  let convertedTemperature;

  switch (this.scale) {
    case 'C':
      convertedTemperature = temperature - 273.15;
      break;
    case 'F':
      convertedTemperature = (temperature * (9 / 5)) - 459.67;
      break;
    default:
      convertedTemperature = temperature;
  }

  return convertedTemperature.toFixed(2);
};

/* Compose the request's URL attaching necessary information. */
OpenWeather.prototype.composeURL = function composeURL(city) {
  return endpoint.replace('{city}', city)
                 .replace('{api}', this.apiKey);
};

module.exports = OpenWeather;


src/openweatherapi.js
// Access to weather stats using OpenWeatherMap APIs.
'use strict';
let http = require('http');

/**!
 * This module allows you to easily retrieve weather using OpenWeather API.
 * All you need is an API key you can get here: https://openweathermap.org/api.
 *
 *   @example
 *     let weather = new OpenWeather();
 *     weather.setAPI('YOUR_API_KEY')
 *     weather.setScale('C');
 *     weather.on('New York');
 */
function OpenWeather() {
    let endpoint = 'http://api.openweathermap.org/data/2.5/weather?q={city}&apikey={api}';
    let api_key  = '';
    let scale = '';

    function setAPI(key) {
        api_key = key;
    }

    function setScale(given_scale) {
        if (!['C', 'F', 'K'].includes(given_scale)) {
            abort("Invalid scale for 'setScale' , expected: C, F or K.");
        }

        scale = given_scale;
    }

    /* Get weather for a specific city. */
    function on(city) {
        let request_url = composeURL(city);

        http.get(request_url, function (res) {
            console.log(`Retrieving weather for '${city.toUpperCase()}'..`);

            res.on('data', formatWeather);
            res.on('error', console.log);
        });
    }

    function formatWeather(data) {
        let suffix = `°${scale}`;

        try {
            data =  JSON.parse(data);

            var weatherStatus = data.weather[0].main;
            var temp = {
                min: convertToUserScale(data.main.temp_min),
                max: convertToUserScale(data.main.temp_max)
            };
        } catch(error) {
            abort('An error occured: check your .howsweather file');
        }

        console.log(`Min: ${temp.min}${suffix}, Max: ${temp.max}${suffix} | ${weatherStatus}`);
        return;
    }

    function convertToUserScale(temperature) {
        switch (scale) {
            case 'C':
              temperature -= 273.15;
              break;
            case 'F':
              temperature = temperature * (9/5) - 459.67;
              break;
        }

        return temperature.toFixed(2);
    }

    /* Compose the request's URL attaching necessary information. */
    function composeURL(city) {
        return endpoint.replace('{city}', city)
                       .replace('{api}', api_key);
    }

    function abort(message) {
        console.log(message);
        process.exit(1);
    }

    return { on: on, setAPI: setAPI, setScale: setScale };
}

module.exports = new OpenWeather();
